Cloud Migration Feature

filipesilva opened the pull request:
This PR includes the following cloud migration endpoints along with the FE user experience for migrating an instance:
```[tasklist]
### Tasks
- [x] `defsetting read-only-mode`
- [x] `metabase.api.cloud-migration` ns
- [x] `cloud_migration` table
- [x] POST `/cloud-migration`
- [x] GET `/cloud-migration`
- [x] PUT `/cloud-migration/cancel`
```

filipesilva committed:
feat: cloud migration endpoints

dpsutton commented on src/metabase/api/cloud_migration.clj:
```diff
@@ -0,0 +1,161 @@
+      (let [cloud-migration (t2/insert-returning-instance! CloudMigration (get-store-migration))]
+        (future (migrate! cloud-migration))
+        cloud-migration)
+      (catch Exception e
```
this `catch` is probably doing less work than you think since the `migrate!` call is in a `future`

filipesilva replied:
It's for the error status codes in `get-store-migration`. TBH I wasn't very happy with putting it here. I should put it in `get-store-migration`, but then I still have somehow say "well if that one returns an error, return that instead".

filipesilva commented on src/metabase/api/cloud_migration.clj:
```diff
@@ -0,0 +1,204 @@
+    :permissions_group_membership})
+
+;; Block write calls to most tables in read-only mode.
+(methodical/defmethod t2.pipeline/build :before [#_query-type     :toucan.statement-type/DML
```
@camsaul PTAL if this is what you had in mind. I used a set instead of a `derive` like you suggested, but maybe I missed why that was important. I'd also appreciate your input on the `read-only-mode-exceptions`.

filipesilva committed:
fix: don't leave open conn behind when checking migrations

`unrun-migrations` would open a new connection but not close it.
Since `with-liquibase` is happy enough using a data-source the fix is straightforward.

You can verify this by running the following code:

```
(comment
  (require '[metabase.cmd.dump-to-h2 :as dump-to-h2]
           '[metabase.analytics.prometheus :as prometheus])

  (defn dump []
    (-> (str "cloud_migration_dump_" (random-uuid) ".mv.db")
        io/file
        .getAbsolutePath
        dump-to-h2/dump-to-h2!))

  (defn busy-conns []
    (-> (prometheus/connection-pool-info)
        first
        second
        :numBusyConnections))

  ;; each dump leaves behind 1 busy conn

  (busy-conns)
  ;; => 0

  (dump)
  (busy-conns)
  ;; => 1

  (dump)
  (busy-conns)
  ;; => 2

  (dump)
  (busy-conns)
  ;; => 3

  )
```

filipesilva commented on src/metabase/db/liquibase.clj:
```diff
@@ -181,7 +181,7 @@
   IMPORTANT: this function takes `data-source` but not `liquibase` because `.listUnrunChangeSets` is buggy. See #38257."
   [^DataSource data-source]
-  (with-liquibase [liquibase (.getConnection data-source)]
+  (with-liquibase [liquibase data-source]
```
See the commit message for details and repro.

qnkhuat replied:
that looks like a problem with `with-liquibase` tho, maybe you can investigate to see where we leave a stay connection unclosed?

Maybe it's not apparent enough but I have the "IMPORTANT" note above there for a reason :(

filipesilva replied:
I think this is the exact place where we leave the conn unclosed. In most other places, `.getConnection` is used with `with-open`, but not here.

I think this fn should either use `with-open`, so the conn is always correctly closed under all circumstances, or it shouldn't open the conn at all and let the called fn do it. I went for the latter option because it looked simpler.

qnkhuat replied:
ah okay, i just realized that `with-liquibase` will open a connection if it takes a  data-source. all good then!

noahmoss commented on src/metabase/api/cloud_migration.clj:
```diff
@@ -0,0 +1,249 @@
+    :model/ViewLog
+
+    ;; TODO: Came up in tests, check with core folks
+    :setting
+    :permissions_group_membership})
```
Not sure why `permissions_group_membership` would need to be writable. Why were tests failing?

`setting` is tricky, because most standard settings probably shouldn't be writable during migration either. Like run-of-the-mill admin settings. If someone updates those while the migration is ongoing, the setting would be lost, right? 

I guess that's also true of the `core_user` table—we'd presumably allow someone to change their password in read-only mode. That might be an issue.

Are we OK from the product POV for the "read-only" mode to be this coarsely grained, and possibly lose data in these cases?

filipesilva replied:
Tested `permissions_group_membership` and `core_user` again, and didn't need it anymore, so removed it. I was getting errors from it at some point, not sure what changed meanwhile.

qnkhuat commented on resources/migrations/001_update_migrations.yaml:
```diff
@@ -6323,6 +6323,64 @@ databaseChangeLog:
+                  name: created_at
+                  remarks: Timestamp when the config was inserted
+                  type: ${timestamp_type}
+                  defaultValueComputed: current_timestamp
```
I'd be careful with `defaultValueComputed`, IIRC this will also add a `ON UPDATE CURRENT_TIMESTAMP` on some MySQL version. To be safe I'd use `(derive :model/CloudMigration :hook/timestamped?)` to let toucan2 handle this.

filipesilva replied:
I have `(derive :hook/timestamped?)` in the model already. So I should just remove `defaultValueComputed` right?

qnkhuat commented on test/metabase/api/cloud_migration_test.clj:
```diff
@@ -0,0 +1,87 @@
+(ns ^:synchronized metabase.api.cloud-migration-test
```
what's the `:synchronized` metadata is for?

filipesilva replied:
To not run the test as parallel. These tests can set the `read-only-mode` setting, which will block DB writes for a lot of different tables. This would interfere with other tests.

qnkhuat replied:
oh in that case I don't think you need the flag because all tests are synchronized by default

filipesilva replied:
Uhm, are you absolutely sure about that? I see other `^:synchronized` tests, and I also see uses of `mt/with-dynamic-redefs` that I'm pretty sure are so that `with-redefs` work with parallel tests (we had this on cloud too at one point).

qnkhuat replied:
well, if you look into the our test runner, it only checks the parallel metadata and it's how it knows which tests should be paralleled: https://github.com/metabase/hawk/blob/490c9b04fbf5861dbd47c0c5281dad2d113bd285/src/mb/hawk/parallel.clj

i don't know where is that synchronized is used at all.

filipesilva replied:
You know, that's a good point. I think I was remembering [eftest/synchronized](https://github.com/weavejester/eftest/blob/5a367d5ebd5fef2062b310d7c4602fdc4785edde/eftest/src/eftest/runner.clj#L22-L24), which hawk is based on. The one I used back at harbormaster was actually `^{:parallel false}`, I even made a [PR](https://github.com/metabase/hawk/commit/a8f3e8a6a53e39357863e33c25506ad7b0028558) to support it at the individual test level.

But neither of those is just `synchronized`. Yet the MB codebase has it in a few tests e.g. [this mongo one](https://github.com/metabase/metabase/blob/585b4ea46b37bdc0d3d5aaafd3377137d21efdee/modules/drivers/mongo/test/metabase/driver/mongo/query_processor_test.clj#L191), and even on a [clj-kondo custom rule](https://github.com/metabase/metabase/blob/585b4ea46b37bdc0d3d5aaafd3377137d21efdee/.clj-kondo/hooks/clojure/test.clj#L135-L137). 

At the same time I see `mt/with-dynamic-redefs` used on tests that don't seem to opt-in to parallelism at all, e.g. [this one](https://github.com/metabase/metabase/blob/585b4ea46b37bdc0d3d5aaafd3377137d21efdee/enterprise/backend/test/metabase_enterprise/upload_management/api_test.clj#L82-L89). I don't see the test runner in deps.edn set parallelism by default either.

So I agree this is a bit of a mystery. I don't want to take on solving this mystery while I'm trying to get this feature in though, so I think the right thing for me to do is remove the metadata like you suggested.

qnkhuat commented on src/metabase/api/cloud_migration.clj:
```diff
@@ -0,0 +1,249 @@
+(defsetting read-only-mode
+  (deferred-tru
+    (str "Boolean indicating whether a Metabase's is in read-only mode with regards to its app db. "
+         "Will take up to 1m to propagate to other Metabase instances in a cluster."
```
do you mean that if you update this setting ,it takes 1m to propagate the change to all the instances?

qnkhuat commented on resources/migrations/001_update_migrations.yaml:
```diff
@@ -6323,6 +6323,64 @@ databaseChangeLog:
             relativeToChangelogFile: true
       rollback: # no rollback necessary, we're not removing the columns yet
 
+  - changeSet:
```
Could you let me know why we need this table? I don't think we need it for coordination; maybe we want to keep a record?

just want to understand the reasoning here.

filipesilva replied:
I actually started this design without the table, thinking an atom would be enough to keep the migration state. But @devurandom correctly brought up this would not work in a cluster environment. We can't control which instance will be running the migration process, and where the endpoint requests are routed.

We don't need the table to coordinate the migration itself (dump+upload), but we need it to coordinate requests from the UI.

qnkhuat replied:
ah, yea that makes sense!

qnkhuat commented on src/metabase/api/cloud_migration.clj:
```diff
@@ -0,0 +1,249 @@
+  nil)
+
+;; Block write calls to most tables in read-only mode.
+(methodical/defmethod t2.pipeline/build :before [#_query-type     :toucan.statement-type/DML
```
Some thoughts:
1. I wonder what alternative we have for intra-instance communication here. Using `settings` does not give you a strong guarantee. Even though we wait for the settings to propagate, but it only propagates if there is a process that "gets" a setting. There are paths where you can create stuff without touching any settings. 
2. have you considered using [table lock](https://www.postgresql.org/docs/current/explicit-locking.html#LOCKING-TABLES)?

qnkhuat reviewed (changes requested):
Hi, it'd help me understand this feature better if there is a product doc or a tech doc linked. But here are some of my initial thoughts on this:
1. I don't think the current mechanism for coordination between instances in a cluster gives us a strong guarantee. I'm interested in whether we can use [table lock](https://www.postgresql.org/docs/current/explicit-locking.html#LOCKING-TABLES) for this.
2. I think the `metabase.api.cloud_migration` namespace has too much code, one way to break this down is to create a `metabase.cloud_migration` namespace to keep all the settings and migration logic.


Edit: found the tech doc: https://www.notion.so/metabase/Tech-Productize-Cloud-migration-030865067e64417a96218588e6f91a72?pvs=4

qnkhuat commented on src/metabase/api/cloud_migration.clj:
```diff
@@ -0,0 +1,249 @@
+  (when (and (read-only-mode)
+             (not (read-only-mode-exceptions model))
+             (not *ignore-read-only-mode*))
+    (throw (ex-info "Metabase is in read-only-mode mode!" {})))
```
nit: this should be i18n'ed, and we probably want a `{:status-code ...}` for the exception data, so the `catch-exceptions` middleware will use that for the response code.

filipesilva replied:
Can you give me an example of what the right way is here? I'll follow it for this one too.

qnkhuat replied:
```clojure
(ex-info (tru "Metabase is in read-only-mode mode!") {:status-code 400 #_or-anything-you-think-is-sane})
```

this should do it

filipesilva replied:
Thank you 🙏

qnkhuat commented on src/metabase/api/cloud_migration.clj:
```diff
@@ -0,0 +1,249 @@
+
+(defn- cluster?
+  []
+  (>= (t2/count 'QRTZ_SCHEDULER_STATE) 2))
```
Interesting! how reliable is this? It'd open a lot of possibilities for us if we can reliably rely on this.

imrkd replied:
I found this an wasn't able to find any situation in which it won't be reliable 🤷

qnkhuat replied:
https://github.com/quartz-scheduler/quartz/issues/733 this is closed without an answer

imrkd replied:
It's an internal quartz mechanism to indicate its running in a cluster. For our use case it should be robust enough, and we didn't want to build a coordination mechanism into metabase for this feature, but honestly it's just a list of active running instances, we could have our own table :)

qnkhuat replied:
yea it'd be nice if we had that mechanism so we don't have this. Our app shouldn't break if Quartz decides to change its mechanism.

filipesilva replied:
@imrkd suggested it. I mostly was looking for something that could make the single node path not have to always wait. I'm happy enough with false negatives sometimes - it should be a strict optimisation over not doing this check.

filipesilva commented:
@qnkhuat added the tech spec to https://github.com/metabase/metabase/issues/39654#issue-2170100160

filipesilva commented:
@qnkhuat I'm not terribly opposed to splitting the ns but I don't see the benefit. It feels like splitting apart things that are indeed together, making them harder to edit and reason about correctly. It also doesn't look out of line for the other api files in the directory.

This isn't my main codebase though. If you feel its important enough it I'll do it.

filipesilva committed:
fix: flush h2 before returning from dump

qnkhuat commented on src/metabase/api/cloud_migration.clj:
```diff
@@ -0,0 +1,245 @@
+
+(def ^:private read-only-mode-exceptions
+  #{;; Migrations need to update their own state
+    :model/CloudMigration
```
I think we should use `table-name` here instead of toucan2 models because you can actually bypass the model and write toucan2 functions that interact with table directly.

```clojure
(t2/insert! :report_card ...)
(t2/insert! (t2/table-name :model/Card) ...)
```

will by pass model and your read only check will not take affect.

But also note that this is not the only perfectly safe. You can actually use honeysql or raw SQL with toucan2.

```clojure
(t2/query "update report_card ...")
(t2/query {:update [:report_card] ...})
```
are both valid

filipesilva replied:
Makes sense, and comparing by table names also lets me remove that "extra" `:setting` exception. Changed.

qnkhuat commented on src/metabase/api/cloud_migration.clj:
```diff
@@ -0,0 +1,245 @@
+
+      (log/info "Dumping h2 backup to" (.getAbsolutePath dump-file))
+      (set-progress id :dump 20)
+      (binding [*ignore-read-only-mode* true]
```
`*ignore-read-only-mode*` is not doing what you think it does here.

all toucan2 operations in `dump-to-h2/dump-to-h2!` are working directly with the table, not the models. so I think if you remove this binding, it'll still work.

See https://github.com/metabase/metabase/pull/41317/files#r1586163568

qnkhuat replied:
Idea: I think you should make `read-only-mode-exceptions` dynamic and dissoc it with models that `dump-to-h2/dump-to-h2!` needs to update during its execution.

filipesilva replied:
Spoke a little bit too soon. The CI tests show the error mode that originally led me to add this var. You can see it [here](https://github.com/metabase/metabase/actions/runs/8910154054/job/24468847652?pr=41317#step:3:1112): a readonly error during key rotation.

Turns out this error doesn't always show up. If you start a repl, call `(dev)`, and run the unit tests without the var, it doesn't show up. But if you call `(do (dev) (start!))` and then call the tests it does show up. I'm not sure exactly how CI is doing it but it does show up there too.

Added it back then.

qnkhuat replied:
Did you remove this after you changed `read-only-mode-exceptions` to a list of table names? if so then it makes sense why the flag works.

wdyt about my idea of making `read-only-mode-exceptions` a dynamic var? then we can remove `*ignore-read-only-mode*`

filipesilva replied:
I had this error before changing the exception to table names, so I don't think that was it.

Regarding making `read-only-mode-exceptions` dynamic... I think that's an option, but I don't feel very inclined to go explore it. It doesn't seem to be an obvious improvement, it's more of a side grade, so it doesn't feel like a great use of my time. I'd have to go change it, then debug it, etc, and I'd end up at the same place.

The core problem is around how the dump key rotation works by binding the current `mdb.connection/*application-db*` to edit the h2 dump. That's why the read-only mode is triggered during dump. But I'm not looking to go fixing dump-to-h2 more than I have to for this feature.

qnkhuat commented on src/metabase/api/cloud_migration.clj:
```diff
@@ -0,0 +1,245 @@
+  (let [dump-file (io/file (str "cloud_migration_dump_" (random-uuid) ".mv.db"))
+        ;; Note: this will still set progress up to twice for each percent
+        ;; because e.g. 70 and 71 upload percent both map to 50+35 migration percent.
+        on-upload-progress #(set-progress id :upload (int (+ 50 (* 50 (/ % 100)))))]
```
why we need to complecate this percentage tho? I'd prefer a simple `#(set-progress id :upload %)`

qnkhuat commented on src/metabase/api/cloud_migration.clj:
```diff
@@ -0,0 +1,245 @@
+        (if (-> e ex-data :terminal)
+          (log/info "Migration interruped due to terminal state")
+          (do
+            (t2/update! CloudMigration :id id {:state :error})
```
```suggestion
            (t2/update! CloudMigration id {:state :error})
```
nit: you don't need this if `id` is the PK already

qnkhuat commented on src/metabase/api/cloud_migration.clj:
```diff
@@ -0,0 +1,245 @@
+  nil)
+
+;; Block write calls to most tables in read-only mode.
+(methodical/defmethod t2.pipeline/build :before [#_query-type     :toucan.statement-type/DML
```
this function should be in `model/interface` IMO. it's where we define all toucan2 methods. 

we should put them together so we know where to look for these kind of things

filipesilva replied:
Tried to but that introduces a cyclic dependency. Pinged you on slack about it.

qnkhuat commented on test/metabase/api/cloud_migration_test.clj:
could you write at test for exceptions model in ready only mode?

At the least we should test that user can still login in read only mode

filipesilva replied:
Added the login test. Saw a warning on setting the last login for the user, it didn't break the test but added the user table to the exceptions as well.

qnkhuat reviewed (commented):
Ok I'll not block this because of how we enter readonly mode, even though I think we should explore using table lock.
I added a comment that we should use table_name, not toucan2 model to be extra sure, please take a look there.

On code organization: I think it's working splitting it up and move some of the things to model/cloud_migration :) API stuffs should only concern about getting and returning things. The business logic should lives in model IMO.

Please take resolve the comment about table/toucan2 model. then I'll approve.
I also think someone from @metabase/admin-webapp should take a look to as this is in their area.

qnkhuat commented:
suggestion for re-organization:
- the defmethod for `t2.pipeline/build` to `metabase.models.interface`
- `migrate!`, all the `defestting`and `terminal-states` to `metabase.models.cloud-migration`

filipesilva committed:
rfct: move code to models.cloud-migration

filipesilva committed:
test: add login while on read-only test

dpsutton commented on src/metabase/models/cloud_migration.clj:
```diff
@@ -0,0 +1,213 @@
+                                                 #_resolved-query :default]
+  [_query-type model _parsed-args resolved-query]
+  (when (and (read-only-mode)
+             (not (read-only-mode-exceptions (t2/table-name model)))
```
why is everything behind `t2/table-name` here? why not just stay in models?

qnkhuat replied:
see : https://github.com/metabase/metabase/pull/41317#discussion_r1586163568, Filipe used model at first.

dpsutton replied:
nice. i think that should have a comment as to why

filipesilva replied:
Added

dpsutton commented on src/metabase/models/cloud_migration.clj:
```diff
@@ -0,0 +1,213 @@
+             (not (read-only-mode-exceptions (t2/table-name model)))
+             (not *ignore-read-only-mode*))
+    (throw (ex-info (tru "Metabase is in read-only-mode mode!")
+                    {:status-code 403})))
```
i get that "you don't have permission to access this resource" kinda makes sense here, but 403 has a pretty standard meaning and throwing it to logged in users feels kinda weird. Is there a better status code for this?

filipesilva replied:
Going down some candidates in https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#client_error_responses:
- 400: the request is fine I think
- 401: client is logged in already
- 403: (the current one) we know the client but they can't do this op (right now)
- 409: this one seems reasonable... it is state related
- 412: there's a precondition missing (read-only false) but not in the headers
- 417: there's a failed expectation (read-only false) but not in the Expect header

So I'd say either 403 or 409. Up to you.

imrkd replied:
403 is intended to be used for logged in users that don't have the authorization to perform a specific operation - `Unlike 401 Unauthorized, the client's identity is known to the server. ` 
I don't think anything else in 4xx/5xx fits.

dpsutton commented on src/metabase/models/cloud_migration.clj:
```diff
@@ -0,0 +1,213 @@
+
+(defn- cluster?
+  []
+  (>= (t2/count 'QRTZ_SCHEDULER_STATE) 2))
```
this seems to work on h2 but on postgres:

```clojure
cloud-migration=> (metabase.db/db-type)
:postgres
cloud-migration=> (cluster?)
Execution error (PSQLException) at org.postgresql.core.v3.QueryExecutorImpl/receiveErrorResponse (QueryExecutorImpl.java:2725).
ERROR: relation "QRTZ_SCHEDULER_STATE" does not exist
  Position: 33
```

Changing this to `(t2/count 'qrtz_scheduler_state)` works on both h2 and postgres. Haven't checked mysql.

filipesilva replied:
Added a test that this should be 1 on tests, that should make the CI raise an error for the engines where it doesn't work.

filipesilva replied:
Looks like it's different just for Postgres, so added a conditional based on db-type
https://github.com/metabase/metabase/blob/0691779ddaf10f6713cdcd528f119c0f6392b056/resources/migrations/000_legacy_migrations.yaml#L4736-L4737

dpsutton commented on src/metabase/models/cloud_migration.clj:
```diff
@@ -0,0 +1,213 @@
+        length                   (.length file)
+        *bytes                   (atom 0)
+        on-percent-progress-memo (memoize on-percent-progress) ;; memoize so we don't repeat calls
+        add-bytes                #(on-percent-progress-memo (int (* 100 (/ (swap! *bytes + %) length))))
```
I'm not sure this on-percent is doing what we think it is doing.

```clojure
cloud-migration=> (let [values (atom [])]
                    (spit "/tmp/input" #_(apply str (repeat 5 "x")) "file contents")
                    (with-open [in (progress-file-input-stream (io/file "/tmp/input") (fn [x] (swap! values conj x)))]
                      (loop []
                        (let [c (.read in)]
                          (when (pos? c)
                            (recur)))))
                    @values)
[784 1592 2423 3200 3446 4207 5061 5907 6800 7576 8423 9315 10200 100]
```

The docstring makes me thing this will send back a progression of `[4 10 ... 96 100]` type numbers, but it's quite different in reality.

dpsutton replied:
it does seem to work correctly for the following:

```clojure
  (into [] (map (fn [size]
                 (let [values (atom [])]
                   (spit "/tmp/input" "file contents")
                   (with-open [in (progress-file-input-stream (io/file "/tmp/input") (fn [x] (swap! values conj x)))]
                     (.read in (byte-array size)))
                   @values)))
        (range 0 17 3))
```

Where i feed it progressively bigger byte arrays to copy into.

filipesilva replied:
Had a bug in `progress-file-input-stream`, expected a char to be returned from arity 1 `.read` but it returns an int. Fixed.

dpsutton commented on src/metabase/models/cloud_migration.clj:
```diff
@@ -0,0 +1,213 @@
+      (read-only-mode! true)
+      (when (cluster?)
+        (log/info "Cluster detected, waiting for read-only mode to propagate")
+        (Thread/sleep (int (* 1.5 setting.cache/cache-update-check-interval-ms))))
```
I think the docstring should say that this should run on a seprate thread due to the heavy sleep here.

filipesilva replied:
Done

qnkhuat commented on src/metabase/models/cloud_migration.clj:
```diff
@@ -0,0 +1,213 @@
+
+(set! *warn-on-reflection* true)
+
+(def CloudMigration "Cloud Migration" :model/CloudMigration)
```
we shouldn't define this for new models, we should just use `:model/CloudMigration` anywhere.

we have this for old models because when we migrate to toucan2 we didnt' want to update every single reference we have in the code base.

filipesilva replied:
Removed

qnkhuat reviewed (approved):
LGTM, but please take a look at Dan's commments.

filipesilva committed:
fix: assorted cloud_migration fixes from review

filipesilva committed:
test: allow overriding uploaded dump

filipesilva committed:
fix: add UserParameterValue to read-only exceptions

Also make the list a little bit nicer.

filipesilva committed:
fix: only block dumped tables on read-only

filipesilva committed:
fix: recover on startup if read-only was left on

filipesilva committed:
feat: block migration when hosted already

filipesilva committed:
test: test settings for migration

filipesilva committed:
feat: cloud migration supports retries and multipart

filipesilva committed:
test: sane dev defaults for migration

filipesilva committed:
fix: upload 100% shouldn't be migration 100%

filipesilva committed:
chore: make up a new migration id after merge

sloansparger committed:
Cloud Migration FE (#42542)

* it's a start

* ui wip

* wip

* dynamic polling intervals, and custom modal for migrate confirmation modal

* cleans out most of the remainig UI TODOs

* adding progress component

* impls team feedback

* makes component more testable, starts some a unit test for the CloudPanel

* finish unit testing

* reverts api changes

* update progress styling

* fix type issues

* fix e2e failure, fix feature unit tests by holding last migration state in fetchMock if more requests than expected happen at the end of a test, remove white spacing change in clj file

* second pass at fixing tests

* fix copy from ready-only to read-only

* copy fix

* Update frontend/src/metabase/admin/settings/components/CloudPanel/MigrationError.tsx

Co-authored-by: Raphael Krut-Landau <raphael.kl@gmail.com>

* Update frontend/src/metabase/admin/settings/components/CloudPanel/MigrationInProgress.tsx

Co-authored-by: Raphael Krut-Landau <raphael.kl@gmail.com>

* adding e2e test

* pr feedback

---------

Co-authored-by: Nick Fitzpatrick <nickfitz.582@gmail.com>
Co-authored-by: Raphael Krut-Landau <raphael.kl@gmail.com>
